spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -407,37 +461,65 @@ private void validateKey(Object entity, PathElement ancestorPE) {
 	 * @return a list of converted entities
 	 */
 	public <T> List<T> convertEntitiesForRead(Iterator<? extends BaseEntity> entities, Class<T> entityClass) {
-		List<T> results = new ArrayList<>();
+		return convertEntitiesForRead(entities, entityClass, new ReadContext());
+	}
+
+	public <T> List<T> convertEntitiesForRead(Iterator<? extends BaseEntity> entities, Class<T> entityClass,
+			ReadContext context) {
 		if (entities == null) {
-			return results;
+			return Collections.emptyList();
 		}
 
 		DatastorePersistentEntity datastorePersistentEntity = this.datastoreMappingContext
 				.getPersistentEntity(entityClass);
 
-		entities.forEachRemaining((entity) -> convertEntityResolveDescendantsAndReferences(
-				entityClass,
-				results, datastorePersistentEntity, entity));
+		List<T> results = StreamSupport
+				.stream(((Iterable<BaseEntity>) () -> (Iterator<BaseEntity>) entities).spliterator(), false)
+				.map((entity) -> convertEntityResolveDescendantsAndReferences(entityClass,
+				datastorePersistentEntity,
+				entity,
+				context))
+				.collect(Collectors.toList());
+
 		return results;
 	}
 
-	private <T> void convertEntityResolveDescendantsAndReferences(Class<T> entityClass,
-			List<T> results, DatastorePersistentEntity datastorePersistentEntity,
-			BaseEntity entity) {
+	private <T> T convertEntityResolveDescendantsAndReferences(Class<T> entityClass,
+			DatastorePersistentEntity datastorePersistentEntity,
+			BaseEntity entity, ReadContext context) {
 		T convertedObject = this.datastoreEntityConverter.read(entityClass, entity);
```

Comment:
Check if entity has a key, and then look it up in the cache first before doing conversion.